api/models/metadata.go

```diff
@@ -0,0 +1,236 @@
+package models
+
+import (
+	"bytes"
+	"database/sql/driver"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+)
+
+// Metadata encapsulates key-value metadata. The structure is immutable via its public API and nil-safe for its contract
+// permissive nilability is here to simplify updates and reduce the need for nil handling in extensions - metadata should be updated by over-writing the original object:
+//  target.Metadata  = target.Metadata.With("fooKey",1)
+// old MD remains empty
+// Metadata is lenable
+type Metadata map[string]*metadataValue
+
+// metadataValue encapsulates a metadata value in the metadata map,
+// This is stored in its un-parsed JSON format for later (re-) parsing into specific structs or values
+// metadataValue objects are  immutable after JSON load
+type metadataValue []byte
+
+const (
+	maxMetadataValueBytes = 512
+	maxMetadataKeyBytes   = 128
+	maxMetadataKeys       = 100
+)
+
+// Equals is defined based on un-ordered k/v comparison at of the metadata keys and  (compacted) values of metadata, JSON object-value equality for values is property-order dependent
+func (m Metadata) Equals(other Metadata) bool {
+	if len(m) != len(other) {
+		return false
+	}
+	for k1, v1 := range m {
+		v2, _ := other[k1]
+		if !bytes.Equal(*v1, *v2) {
+			return false
+		}
+	}
+	return true
+}
+
+func EmptyMetadata() Metadata {
+	return nil
+}
+
+func (mv *metadataValue) String() string {
+	return string(*mv)
+}
+
+func (v *metadataValue) MarshalJSON() ([]byte, error) {
+	return *v, nil
+}
+
+func (mv *metadataValue) isEmptyValue() bool {
+	sval := string(*mv)
+	return sval == "\"\"" || sval == "null"
+}
+
+// UnmarshalJSON compacts metadata values but does not alter key-ordering for keys
+func (mv *metadataValue) UnmarshalJSON(val []byte) error {
+	buf := bytes.Buffer{}
+	err := json.Compact(&buf, val)
```

Comment:
i'm not sure we should do this, specifically this way at least. nothing against the minification movement, it seems nice in theory, but it may change input for configs where spacing matters (yaml) and that kind of data seems a likely possibility in this field :( - the easy thing to do seems to just be validate `val` is utf8 and `*mv = val` ?

Reply:
Hmm maybe I’ve got something wrong here, 

I assumed that for unmarshal here the byte slice is the json sub-doc corresponding to the value (in the metadata map) (i.e quoted string, array object, number), and that this byte slice is a sub-slice of the original input buffer (from the user)

My reading of the json.Compact docs was that it preserves the content model   (property ordering, string content),and that, given the metadata map is defined as JSON in the request, that whitespace was ignorable

So, the compact here is just to remove any superfluous whitespav which might be present in the metadata sub-doc from API requests , in order to get a consistent value for the (JSON) value size, I don’t think there is any way that it changes the meaning of metadata values themselves

Reply:
it at least appears to skip over `\r\n` https://golang.org/src/encoding/json/indent.go?s=284:333#L1 -- I could be wrong, didn't thoroughly read their json parser code. I tinkered with the playground for a minute and couldn't get an example to not error, either I'm completely missing what a metadata value is or json compact isn't gonna work so great for any values that have a newline: https://play.golang.org/p/U7vqPeoob0Q

Reply:
json.Compact never touches values, only interstitial whitespace between JSON tokens - the specific issue here would be :

User makes API request using indented JSON (not that they likely ever would) - the whole API request is indented: 

```
POST /v1/apps 

{
     "app" { 
           ...
          "metadata" : {
                       "md.key.1": {  
                                   "a": v1 \n with \n newlines    and arbitrary spacing ", 

                                   "b" : 10
                        }, 
                       "md.key.2": "v2"
           }
          ....
}
```

In this case the extracted Metadata doc for key "md.key.1" (passed to metadataValue.UnmarshalJSON) 
```
 {  
                                   "a": v1 \n with \n newlines    and arbitrary spacing ", 

                                   "b" : 10
                        }
```

If I removed the `json.Compact` that would be what we stored in the DB  verbatim, and that would then be was returned in subsequent API responses containing that metadata. 

I'm assuminging that (from a user's perspetive) any interstitial whitespace within the metadata values  is non-semantic (as it is on the API request that sets the metadata itself)

Stripping the whitespace within the value sub-doc avoids any confusion about storage sizes  (an indented request and a semantically identical compact request get the same size limits), and also consolidates equality (changing the spacing of your API requests doesn't change equality w.r.t updates). 

Equality is of course a little bit tricky here due to key order preservation, I opted for preserving the original key ordering within values (hence keeping the values as unparsed byte slices) as I thought that was least-surprise for consumers.


If I'd stored metadata as  `map[string]interface{}` the  compaction would happen via the internal unmarshal/marshal in the same way as with json.Compact, however Equality would then be order-independent for value maps and more importantly extensions would need to render that value interface{} back int to a byte array in order to then re-parse it into their own structs. 

I added a quick equality check to validate the correct behaviour. 

`{mustParseMd(t,

			`{
 "md.1" : {
           
   "subkey1": "value\n with\n newlines",
   "subkey2": true 

}
}`), mustParseMd(t,`{"md.1":{"subkey1":"value\n with\n newlines","subkey2":true}}`), true},`